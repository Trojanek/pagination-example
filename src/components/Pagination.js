import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import Button from '@material-ui/core/Button';
import IconButton from '@material-ui/core/IconButton';
import PrevIcon from '@material-ui/icons/NavigateBefore';
import NextIcon from '@material-ui/icons/NavigateNext';

const navigations = [-1, 0, 1];

const styles = () => ({
    root: {
        display: 'flex',
        flexDirection: 'row',
        alignItems: 'center',
        justifyContent: 'space-between',
        width: 300,
    },
});

const Pagination = ({
    itemsCount,
    itemsPerPage,
    currentPage,
    onChange,
    classes,
}) => {
    const minPage = 1;
    const maxPage = Math.ceil((itemsCount || 1) / itemsPerPage);
    const hasPrev = currentPage > minPage;
    const hasNext = currentPage < maxPage;
    const renderNavigations = item => (
        (currentPage + item >= minPage && currentPage + item <= maxPage)
            ? (
                <Button
                    key={item}
                    onClick={() => onChange(currentPage + item)}
                    disabled={item === 0}
                >
                    {currentPage + item}
                </Button>
            )
            : null
    );
    return (
        <div className={classes.root}>
            <IconButton
                onClick={() => onChange(currentPage - 1)}
                disabled={!hasPrev}
                aria-label="Previous Page"
            >
                <PrevIcon />
            </IconButton>
            <div className={classes.navigations}>
                {navigations.map(renderNavigations)}
            </div>
            <IconButton
                onClick={() => onChange(currentPage + 1)}
                disabled={!hasNext}
                aria-label="Next Page"
            >
                <NextIcon />
            </IconButton>
        </div>
    );
};

Pagination.propTypes = {
    itemsCount: PropTypes.number.isRequired,
    itemsPerPage: PropTypes.number.isRequired,
    currentPage: PropTypes.number.isRequired,
    onChange: PropTypes.func.isRequired,
    classes: PropTypes.object.isRequired,
};

export default withStyles(styles)(Pagination);
