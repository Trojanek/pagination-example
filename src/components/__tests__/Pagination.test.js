import React from 'react';
import { shallow } from 'enzyme';
import Pagination from '../Pagination';

const onChange = jest.fn();


describe('<Pagination />', () => {
    let pagination;

    beforeEach(() => {
        pagination = shallow((
            <Pagination
                itemsCount={41}
                itemsPerPage={10}
                currentPage={1}
                onChange={onChange}
            />
        )).first().shallow();
    });

    it('should render correctly', () => {
        expect(pagination).toMatchSnapshot();
    });

    it('should render correctly with 0 items', () => {
        pagination.setProps({
            itemsCount: 0,
        });
        expect(pagination).toMatchSnapshot();
    });

    it('should render correctly last page', () => {
        pagination.setProps({
            currentPage: 5,
        });
        expect(pagination).toMatchSnapshot();
    });

    it('should render disabled prev button on first page', () => {
        const received = pagination
            .find('WithStyles(IconButton)[aria-label="Previous Page"]')
            .at(0)
            .prop('disabled');
        expect(received).toBe(true);
    });

    it('should render enabled next button on first page', () => {
        const received = pagination
            .find('WithStyles(IconButton)[aria-label="Next Page"]')
            .at(0)
            .prop('disabled');
        expect(received).toBe(false);
    });

    it('should render enabled prev button on last page', () => {
        pagination.setProps({
            currentPage: 5,
        });
        const received = pagination
            .find('WithStyles(IconButton)[aria-label="Previous Page"]')
            .at(0)
            .prop('disabled');
        expect(received).toBe(false);
    });

    it('should render disabled next button on last page', () => {
        pagination.setProps({
            currentPage: 5,
        });
        const received = pagination
            .find('WithStyles(IconButton)[aria-label="Next Page"]')
            .at(0)
            .prop('disabled');
        expect(received).toBe(true);
    });

    it('should call "onChage" with proper argument on second page', () => {
        const expected = 1;
        const callsCount = onChange.mock.calls.length;
        pagination.setProps({
            currentPage: 2,
        });
        pagination
            .find('WithStyles(IconButton)[aria-label="Previous Page"]')
            .at(0)
            .simulate('click');
        expect(onChange).toHaveBeenCalledTimes(callsCount + 1);
        expect(onChange).toHaveBeenLastCalledWith(expected);
    });

    it('should call "onChage" after click "next" with proper argument 4th page', () => {
        const expected = 5;
        const callsCount = onChange.mock.calls.length;
        pagination.setProps({
            currentPage: 4,
        });
        pagination
            .find('WithStyles(IconButton)[aria-label="Next Page"]')
            .at(0)
            .simulate('click');
        expect(onChange).toHaveBeenCalledTimes(callsCount + 1);
        expect(onChange).toHaveBeenLastCalledWith(expected);
    });

    it('should call "onChage" with proper argument on first page', () => {
        const expected = 2;
        const callsCount = onChange.mock.calls.length;
        pagination
            .find('WithStyles(Button)')
            .at(1)
            .simulate('click');
        expect(onChange).toHaveBeenCalledTimes(callsCount + 1);
        expect(onChange).toHaveBeenLastCalledWith(expected);
    });
});
