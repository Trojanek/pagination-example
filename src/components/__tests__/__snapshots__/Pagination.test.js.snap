// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> should render correctly 1`] = `
<div
  className="Pagination-root-1"
>
  <WithStyles(IconButton)
    aria-label="Previous Page"
    disabled={true}
    onClick={[Function]}
  >
    <pure(NavigateBeforeIcon) />
  </WithStyles(IconButton)>
  <div>
    <WithStyles(Button)
      disabled={true}
      key="0"
      onClick={[Function]}
    >
      1
    </WithStyles(Button)>
    <WithStyles(Button)
      disabled={false}
      key="1"
      onClick={[Function]}
    >
      2
    </WithStyles(Button)>
  </div>
  <WithStyles(IconButton)
    aria-label="Next Page"
    disabled={false}
    onClick={[Function]}
  >
    <pure(NavigateNextIcon) />
  </WithStyles(IconButton)>
</div>
`;

exports[`<Pagination /> should render correctly last page 1`] = `
<div
  className="Pagination-root-1"
>
  <WithStyles(IconButton)
    aria-label="Previous Page"
    disabled={false}
    onClick={[Function]}
  >
    <pure(NavigateBeforeIcon) />
  </WithStyles(IconButton)>
  <div>
    <WithStyles(Button)
      disabled={false}
      key="-1"
      onClick={[Function]}
    >
      4
    </WithStyles(Button)>
    <WithStyles(Button)
      disabled={true}
      key="0"
      onClick={[Function]}
    >
      5
    </WithStyles(Button)>
  </div>
  <WithStyles(IconButton)
    aria-label="Next Page"
    disabled={true}
    onClick={[Function]}
  >
    <pure(NavigateNextIcon) />
  </WithStyles(IconButton)>
</div>
`;

exports[`<Pagination /> should render correctly with 0 items 1`] = `
<div
  className="Pagination-root-1"
>
  <WithStyles(IconButton)
    aria-label="Previous Page"
    disabled={true}
    onClick={[Function]}
  >
    <pure(NavigateBeforeIcon) />
  </WithStyles(IconButton)>
  <div>
    <WithStyles(Button)
      disabled={true}
      key="0"
      onClick={[Function]}
    >
      1
    </WithStyles(Button)>
  </div>
  <WithStyles(IconButton)
    aria-label="Next Page"
    disabled={true}
    onClick={[Function]}
  >
    <pure(NavigateNextIcon) />
  </WithStyles(IconButton)>
</div>
`;
